core/domain/topic_domain.py

```diff
@@ -133,6 +157,57 @@ def to_dict(self):
             'version': self.version
         }
 
+    def validate(self):
+        """Validates all properties of this topic and its constituents.
+
+        Raises:
+            ValidationError: One or more attributes of the Topic are not
+            valid.
+        """
+        if not isinstance(self.name, basestring):
+            raise utils.ValidationError(
+                'Expected name to be a string, received %s' % self.name)
+        if not isinstance(self.description, basestring):
+            raise utils.ValidationError(
+                'Expected description to be a string, received %s'
+                % self.description)
+
+        if not isinstance(self.language_code, basestring):
+            raise utils.ValidationError(
+                'Expected language code to be a string, received %s' %
+                self.language_code)
+        if not any([self.language_code == lc['code']
+                    for lc in constants.ALL_LANGUAGE_CODES]):
+            raise utils.ValidationError(
+                'Invalid language code: %s' % self.language_code)
+
+        if not isinstance(self.canonical_story_ids, list):
+            raise utils.ValidationError(
+                'Expected canonical story ids to be a list, received %s'
+                % self.canonical_story_ids)
+        for story_id in self.canonical_story_ids:
+            if story_services.get_story_by_id(story_id, False) is None:
```

Comment:
When passing args with default values, always include the argname: get_story_by_id(story_id, strict=False).

That said -- domain object modules, in general, should not depend on services. I think it is OK, here, to just check that it's a list of strings. You don't need to actually fetch the stories for every commit -- that's rather heavy.

As a sanity-check, I noticed that collection_domain.py does not import exp_services. Do you know where (if at all) they check that the explorations making up the collection are valid?

Reply:
In collection, in _save_collection(), it is done. Shall I move these type of validations just before saving then, in all the constructs?

Reply:
Done, removed those checks.

Reply:
Ah ok, thanks for the info.

Let me bring this up with the other tech leads -- I'm not sure what the correct thing to do is, especially since there are a lot more cases here of things depending on things and it becomes a question of architecture. I'll get some thoughts and write back to you!

Reply:
Ok, thanks!